### PR TITLE
[WIP] lightning_base: support --lr_scheduler with multiple possibilities

### PR DESCRIPTION
**Pull request by stas00**
This is step 1 to implement https://github.com/huggingface/transformers/issues/6070

This PR adds support to:

```
python finetune.py [...]  --lr_scheduler=SCHEDULER [...]
```

in lightning_base (so seq2seq examples, etc.)

to get help:

```
python finetune.py [...]  --lr_scheduler=help [...]
```
gives:
```
Available lr_schedulers:
--lr_scheduler=cosine (get_cosine_schedule_with_warmup)
--lr_scheduler=cosine_w_restarts (get_cosine_with_hard_restarts_schedule_with_warmup)
--lr_scheduler=linear (get_linear_schedule_with_warmup)
--lr_scheduler=help (this help)
```
the rest is hopefully obvious.

I'm not sure where to document this new option - I added a note to `seq2seq/README.md`, but it should work for any other module under examples/. That makes me think that in this PR https://github.com/huggingface/transformers/pull/6149 the new options are only documented in `seq2seq/README.md`, but should be usable under other examples - i.e. it's not s2s-specific, but examples/README.md doesn't look like it's going into any details. Thoughts?

I wrote tests inside `seq2seq` too, but it should work anywhere.

@sshleifer

### CODE REVIEW

File: examples/lightning_base.py [hunk 1]
```diff
@@ -39,6 +40,18 @@
 }
 
 
+# update this to support new schedulers from transformers.optimization
+# the listed below schedulers appear in the output of --lr_scheduler=help
+arg_to_scheduler = {
+    "linear": "get_linear_schedule_with_warmup",
+    "cosine": "get_cosine_schedule_with_warmup",
+    "cosine_w_restarts": "get_cosine_with_hard_restarts_schedule_with_warmup",
```
Comment (sshleifer) [inline examples/lightning_base.py:48]: consider importing the functions you use and then hardcoding
```python
arg_to_scheduler = {"linear": get_linear_schedule_with_warmup, ...}
```
That way IDEs can find usages.

File: examples/lightning_base.py [hunk 2]
```diff
@@ -206,6 +234,12 @@ def add_model_specific_args(parser, root_dir):
             "--attention_dropout", type=float, help="Attention dropout probability (Optional). Goes into model.config",
         )
         parser.add_argument("--learning_rate", default=5e-5, type=float, help="The initial learning rate for Adam.")
+        parser.add_argument(
+            "--lr_scheduler",
+            default="linear",
+            type=str,
```
Comment (sshleifer) [inline examples/lightning_base.py:240]: if you pass `choices=["linear", "cosine", "cosine_with_restarts"]` I think you will get a nice traceback for `--lr_scheduler=help` without having to add an override the "help" value.

**Review by sshleifer — COMMENTED** [review]
Awesome, left a few nits.

File: examples/lightning_base.py [hunk 3]
```diff
@@ -71,6 +84,15 @@ def __init__(
...
             self.config: PretrainedConfig = config
 
+        # usage extended helpers
+        lr_scheduler = hparams.lr_scheduler
+        if lr_scheduler not in arg_to_scheduler.keys():
+            if lr_scheduler == "help":
+                print(lr_schedulers_help())
+                sys.exit(0)
+            else:
+                assert False, f"Error: Invalid --lr_scheduler option: {lr_scheduler}\n" + lr_schedulers_help()
```
Comment (sshleifer) [inline examples/lightning_base.py:94]: I think `raise InvalidArgumentError(msg)` is more pythonic

File: examples/lightning_base.py (hunk 1 above)
Comment (sgugger) [reply examples/lightning_base.py:48]: +1 on this

**Review by sgugger — APPROVED** [review]
Looks good, thanks!

File: examples/lightning_base.py (hunk 1 above)
Comment (stas00) [reply examples/lightning_base.py:48]: I used the single source approach, to avoid maintaining multiple entries of the same. So that the import happens on demand.

With your suggestion, we need to repeat each entry twice - import + populate dict (and to do =help, need to derive the function name from the object, but oh well, that's easy).

An alternative solution is to import `transformers.optimization` and then use the full form in the dic:
```
import transformers.optimization
arg_to_scheduler = {
    "linear": transformers.optimization.get_linear_schedule_with_warmup,
    "cosine":  transformers.optimization.get_cosine_schedule_with_warmup,
    ...
```

File: examples/lightning_base.py (hunk 2 above)
Comment (stas00) [reply examples/lightning_base.py:240]: yes, when we originally discussed this - that was your other suggestion and your said the =help override is preferable. But I will change it to this suggestion. Thank you.

Comment (LysandreJik) [reply examples/lightning_base.py:240]: I second this, putting them in `choices` means that when doing `python script.py --help`, the choices will be shown for the `lr_scheduler` argument. This way there's no need to handle  the help and invalid scheduler option.

**Review by LysandreJik — APPROVED** [review]
Nice, very cool to have added a test!

File: examples/lightning_base.py (hunk 2 above)
Comment (stas00) [reply examples/lightning_base.py:240]: Ah, one shortcoming of this approach is that it'll not be able to tell the user which actual scheduler will be run, just the alias.
Current code produces:
```
Available lr_schedulers:
--lr_scheduler=cosine (get_cosine_schedule_with_warmup)
--lr_scheduler=cosine_w_restarts (get_cosine_with_hard_restarts_schedule_with_warmup)
--lr_scheduler=linear (get_linear_schedule_with_warmup)
--lr_scheduler=help (this help)
```
choices will just list the keys. 

thoughts?

Comment (sgugger) [reply examples/lightning_base.py:240]: I think the aliases are clear enough, no?

Comment (stas00) [reply examples/lightning_base.py:240]: Should there also be some sort of logger that optinally logs which scheduler is being run? Or is it unnecessary?

**Comment by stas00** [comment]
I noticed another thing, `seq2seq/finetune.py`, has a hardcoded:

```
        scheduler = get_linear_schedule_with_warmup(
            self.opt, num_warmup_steps=self.hparams.warmup_steps, num_training_steps=t_total
        )
```

should I switch to using a superclass' `get_lr_scheduler` method that this PR ads instead ?

I haven't checked the other sub-classes - probably needs the same. Or leave that to another PR?

File: examples/lightning_base.py (hunk 2 above)
Comment (stas00) [reply examples/lightning_base.py:240]: ok, using `choices` results in an ugly output, e.g.:

```
                   [--learning_rate LEARNING_RATE]
                   [--lr_scheduler {aaa,bbb,cccaaa,bbb,cccaaa,bbb,cccaaa,bbb,cccaaa,bbb,cccaaa,bbb,cccaaa,bbb,cccaaa,bbb,cccaaa,bbb,cccaaa,bbb,cccaaa,bbb,ccc}]
                   [--weight_decay WEIGHT_DECAY] [--adam_epsilon ADAM_EPSILON]
```

**Comment by stas00** [comment]
All the requested changes have been addressed. 

Currently we get:

```
finetune.py [...] --lr_scheduler=cosine1
finetune.py: error: argument --lr_scheduler: invalid choice: 'cosine1' (choose from 'cosine', 'cosine_w_restarts', 'linear')
```


```
finetune.py [...] --help
[...]
                   [--learning_rate LEARNING_RATE]
                   [--lr_scheduler {cosine, cosine_w_restarts, linear}]
                   [--weight_decay WEIGHT_DECAY] [--adam_epsilon ADAM_EPSILON]
[...]
```
I used `metavar`  in addition to `choices` to make the output a bit user-friendlier - it'll still be less so once there will be many schedulers on that list - `argparse` doesn't wrap them and doesn't allow multiline `metavar`.

**Review by LysandreJik — APPROVED** [review]
I think this is sufficient as self-documentation. Ok to merge for me!
